test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+
+        with self.assertRaises(AttributeError) as context:
+            db._does_not_exist
+
+        # Message should be: "AttributeError: Database has no attribute
+        # '_does_not_exist'. To access the _does_not_exist collection,
+        # use database['_does_not_exist']".
+        self.assertIn("has no attribute '_does_not_exist'", str(context.exception))
+
+    def test_iteration(self):
+        db = self.client.pymongo_test
+        if "PyPy" in sys.version and sys.version_info < (3, 8, 15):
+            msg = "'NoneType' object is not callable"
+        else:
+            if _IS_SYNC:
+                msg = "'Database' object is not iterable"
+            else:
+                msg = "'AsyncDatabase' object is not iterable"
+        # Iteration fails
+        with self.assertRaisesRegex(TypeError, msg):
+            for _ in db:  # type: ignore[misc] # error: "None" not callable  [misc]
+                break
+        # Index fails
+        with self.assertRaises(TypeError):
+            _ = db[0]
+        # next fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = next(db)
+        # .next() fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = db.next()
+        # Do not implement typing.Iterable.
+        self.assertNotIsInstance(db, Iterable)
+
+
+class TestDatabase(AsyncIntegrationTest):
+    async def test_equality(self):
+        self.assertNotEqual(AsyncDatabase(self.client, "test"), AsyncDatabase(self.client, "mike"))
+        self.assertEqual(AsyncDatabase(self.client, "test"), AsyncDatabase(self.client, "test"))
+
+        # Explicitly test inequality
+        self.assertFalse(AsyncDatabase(self.client, "test") != AsyncDatabase(self.client, "test"))
+
+    async def test_hashable(self):
+        self.assertIn(self.client.test, {AsyncDatabase(self.client, "test")})
+
+    async def test_get_coll(self):
+        db = AsyncDatabase(self.client, "pymongo_test")
+        self.assertEqual(db.test, db["test"])
+        self.assertEqual(db.test, AsyncCollection(db, "test"))
+        self.assertNotEqual(db.test, AsyncCollection(db, "mike"))
+        self.assertEqual(db.test.mike, db["test.mike"])
+
+    async def test_repr(self):
+        if _IS_SYNC:
+            name = "Database"
+        else:
+            name = "AsyncDatabase"
+        self.assertEqual(
+            repr(AsyncDatabase(self.client, "pymongo_test")),
+            "{}({!r}, {})".format(name, self.client, repr("pymongo_test")),
+        )
+
+    async def test_create_collection(self):
+        db = AsyncDatabase(self.client, "pymongo_test")
+
+        await db.test.insert_one({"hello": "world"})
+        with self.assertRaises(CollectionInvalid):
+            await db.create_collection("test")
+
+        await db.drop_collection("test")
+
+        with self.assertRaises(TypeError):
+            await db.create_collection(5)
```

Comment:
The switch from 

```
self.assertRaises(TypeError, db.create_collection, 5)
```

to 

```
with self.assertRaises(TypeError):
    await db.create_collection(5)
```

Will cause type errors due to the `db.create_collection(5)` call being explicit instead of done for us by the `assertRaises(...)` call.

All of the typechecking failures that look like `Argument 1 to "create_collection" of "AsyncDatabase" has incompatible type "None"` can be fixed by adding `# type: ignore[arg-type]` at the end of the line, similar to what's done on line 178 here.